### PR TITLE
storage_proxy: Include counter writes in `writes_coordinator_outside_replica_set`

### LINKED ISSUE #4337: storage_proxy counters are not updated on cql counter operations
Version : Scylla 3.0.3

reproduce 

```
cassandra-stress counter_write no-warmup n=1000 -node 172.17.0.2

curl http://172.17.0.2:9180/metrics | grep write_attempts
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP scylla_storage_proxy_coordinator_read_repair_write_attempts_local_node number of write operations in a read repair contexton a local Node
# TYPE scylla_storage_proxy_coordinator_read_repair_write_attempts_local_node counter
scylla_storage_proxy_coordinator_read_repair_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="0",type="derive"} 0
scylla_storage_proxy_coordinator_read_repair_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="1",type="derive"} 0
scylla_storage_proxy_coordinator_read_repair_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="2",type="derive"} 0
scylla_storage_proxy_coordinator_read_repair_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="3",type="derive"} 0
# HELP scylla_storage_proxy_coordinator_total_write_attempts_local_node total number of write requestson a local Node
# TYPE scylla_storage_proxy_coordinator_total_write_attempts_local_node counter
scylla_storage_proxy_coordinator_total_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="0",type="derive"} 38
scylla_storage_proxy_coordinator_total_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="1",type="derive"} 11
scylla_storage_proxy_coordinator_total_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="2",type="derive"} 7
scylla_storage_proxy_coordinator_total_write_attempts_local_node{instance="e484ce6da61b",op_type="mutation_data",shard="3",type="derive"} 7
100  217k    0  217k    0     0  40.3M      0 --:--:-- --:--:-- --:--:-- 42.4M
[root@e484ce6da61b /]# curl http://172.17.0.2:9180/metrics | grep cql_update
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP scylla_cql_updates Counts a total number of CQL UPDATE requests.
# TYPE scylla_cql_updates counter
scylla_cql_updates{instance="e484ce6da61b",shard="0",type="derive"} 211
scylla_cql_updates{instance="e484ce6da61b",shard="1",type="derive"} 260
scylla_cql_updates{instance="e484ce6da61b",shard="2",type="derive"} 249
scylla_cql_updates{instance="e484ce6da61b",shard="3",type="derive"} 308
100  217k    0  217k    0     0  35.3M      0 --:--:-- --:--:-- --:--:-- 42.4M
[root@e484ce6da61b /]# curl http://172.17.0.2:9180/metrics | grep cql_insert
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  8058    0  8058    0     0  4548k      0 --:--:-- --:--:-- --:--:-- 3934k# HELP scylla_cql_inserts Counts a total number of CQL INSERT requests.
# TYPE scylla_cql_inserts counter
scylla_cql_inserts{instance="e484ce6da61b",shard="0",type="derive"} 8
scylla_cql_inserts{instance="e484ce6da61b",shard="1",type="derive"} 4
scylla_cql_inserts{instance="e484ce6da61b",shard="2",type="derive"} 0
scylla_cql_inserts{instance="e484ce6da61b",shard="3",type="derive"} 0
100  217k    0  217k    0     0  20.8M      0 --:--:-- --:--:-- --:--:-- 21.2M
```

we are doing 1000 CQL updates to counters  and thos are counted in ```scylla_cql_updates``` but the ```scylla_storage_proxy_coordinator_total_write_attempts_local_node``` is not updated

### PR DESCRIPTION
**Pull request by jul-stas**
With this change, coordinator prefers himself as the "counter leader", so if another endpoint is chosen as the leader, we know that coordinator was not a member of replica set. With this guarantee we can increment `scylla_storage_proxy_coordinator_writes_coordinator_outside_replica_set` metric after electing different leader (that metric used to neglect the counter updates).

The motivation for this change is to have more reliable way of counting non-token-aware queries.

Fixes #4337

### CODE REVIEW

File: service/storage_proxy.cc [hunk 1]
```diff
@@ -2141,16 +2141,24 @@ gms::inet_address storage_proxy::find_leader_for_counter_update(const mutation&
         throw exceptions::unavailable_exception(cl, block_for(ks, cl), 0);
     }
 
+    const auto my_address = utils::fb_utilities::get_broadcast_address();
+    bool i_can_be_leader = false;
     auto local_endpoints = boost::copy_range<std::vector<gms::inet_address>>(live_endpoints | boost::adaptors::filtered([&] (auto&& ep) {
+        i_can_be_leader |= ep == my_address;
```
Comment (gleb-cloudius) [inline service/storage_proxy.cc:2147]: I personally do not like abusing copy_range like that. I think it should not be a problem to have additional pass over local_endpoints to look for my_address. But I know that some performance paranoid people may not agree with that.

**Comment by gleb-cloudius** [comment]
I do not mind the first patch, but it is not really needed to update the counter. It can be updated directly in `find_leader_for_counter_update` without changing leader selection logic there.

**Comment by jul-stas** [comment]
> I do not mind the first patch, but it is not really needed to update the counter. It can be updated directly in `find_leader_for_counter_update` without changing leader selection logic there.

True; but that should have been done long ago, so I took the opportunity.
Also, this way - instead of doing `++`s on metric - we can do less `+=`s :smiley:

**Review by psarna — APPROVED** [review]

File: service/storage_proxy.cc (hunk 1 above)
Comment (jul-stas) [reply service/storage_proxy.cc]: ~Done. Most clients are token-aware (I think) so I put the short circuit even before the creation of `local_endpoints`.

File: service/storage_proxy.cc [hunk 2]
```diff
@@ -2141,16 +2141,24 @@ gms::inet_address storage_proxy::find_leader_for_counter_update(const mutation&
         throw exceptions::unavailable_exception(cl, block_for(ks, cl), 0);
     }
 
+    const auto my_address = utils::fb_utilities::get_broadcast_address();
+    for (const auto& ep : live_endpoints) {
+        // Favour ourselves to avoid additional hop
```
Comment (gleb-cloudius) [inline service/storage_proxy.cc:2146]: You do no need to search in `live_endpoints`, but in `local_endpoint` and you can do it with `boost::find`.

File: service/storage_proxy.cc [hunk 3]
```diff
@@ -2141,16 +2141,24 @@ gms::inet_address storage_proxy::find_leader_for_counter_update(const mutation&
...
     }));
+
     if (local_endpoints.empty()) {
         // FIXME: O(n log n) to get maximum
         auto& snitch = locator::i_endpoint_snitch::get_local_snitch_ptr();
-        snitch->sort_by_proximity(utils::fb_utilities::get_broadcast_address(), live_endpoints);
+        snitch->sort_by_proximity(my_address, live_endpoints);
         return live_endpoints[0];
     } else {
-        // FIXME: favour ourselves to avoid additional hop?
```
Comment (gleb-cloudius) [inline service/storage_proxy.cc:2153]: do it here.

File: service/storage_proxy.cc (hunk 1 above)
Comment (gleb-cloudius) [reply service/storage_proxy.cc]: Ah, so you did it deliberately? Well, use boost::find then and add a comment that you do it deliberately. Or just drop the use of `boost::copy_range` and open code the loop that will return immediately if local address is found.

File: service/storage_proxy.cc (hunk 2 above)
Comment (jul-stas) [reply service/storage_proxy.cc:2146]: Agreed with `boost::find`, but I think that (with modern drivers) this early return will fire almost always. So by exiting before the creation of `local_endpoints` we save a vector copy at the cost of iterating longer table.

Anyway, it's not worth that much discussion. If you are not convinced with the above please just say "no".

Comment (gleb-cloudius) [reply service/storage_proxy.cc:2146]: See my other comment here: https://github.com/scylladb/scylla/pull/8282#discussion_r595048611

File: service/storage_proxy.cc (hunk 1 above)
Comment (jul-stas) [reply service/storage_proxy.cc]: Done. I chose the "early `boost::find` path" + comment.

**Review by gleb-cloudius — APPROVED** [review]

**Pull request merged**